[benchmarks] Add array access microbenchmark (#132)
google/safeside: benchmarks/spectre_v1/CMakeLists.txt

cmake_minimum_required(VERSION 3.5.1)
set(CMAKE_BUILD_TYPE Release)

project(safeside VERSION 0.1.0 LANGUAGES CXX)

# Use C++11 without extensions
set(CMAKE_CXX_STANDARD 11)

> set(CMAKE_CXX_EXTENSIONS off)

Reviewer: These only need to be configured once for the project; you can leave them out here.

add_executable(unmitigated microbenchmark.cc)

> target_compile_options(unmitigated PUBLIC -O0)

Reviewer: We can probably leave out optimization flags; they're provided at toplevel based on the kind of build. The important thing here is that the *same* flags are used for the unmitigated and mitigated versions, which should be true as long as we don't override them.

Reviewer: Ah, nevermind, I think I understand -- you're using `-O0` to prevent the compiler from optimizing away the entire benchmark, since the array manipulation doesn't actually have visible effects.

That's okay for now, but please leave a comment explaining what we're up to.

Later, we can replace this with tricks like [forcing the compiler to compute values eagerly](https://github.com/travisdowns/zero-fill-bench/blob/c30d8fc88d3f560ba165ef756167a2a703678134/opt-control.hpp#L11-L36).

Author: I added the comment and FIXME. Would it be better to add a github issue for this? What's the convention?

Reviewer: FIXME seems fine.

> if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")

Reviewer: ```suggestion
if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
```

for annoying reasons described in https://github.com/google/safeside/commit/9ca72f2dc0dbe8bb182bca534dd417c6fb6e558e

> 	add_executable(slh_mitigated microbenchmark.cc)
> 	target_compile_options(slh_mitigated PUBLIC -mspeculative-load-hardening -O0)

Reviewer: ```suggestion
  add_executable(slh_mitigated microbenchmark.cc)
  target_compile_options(slh_mitigated PUBLIC -mspeculative-load-hardening -O0)
```

Build files in the repo indent with two spaces, or four in the case of continued lines.

add_executable(unmitigated microbenchmark.cc)
# Compile with -O0 so the array manipulation doesn't get optimized away by the
# compiler.
# FIXME: Replace this with tricks like this:
# https://github.com/travisdowns/zero-fill-bench/blob/c30d8fc88d3f560ba165ef756167a2a703678134/opt-control.hpp#L11-L36
target_compile_options(unmitigated PUBLIC -O0)

> if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
>   # The Clang flag -mspeculative-load-hardening isn't available until Clang
>   # 8.0.0.
>   if("${CMAKE_CXX_COMPILER_VERSION_VERSION}" VERSION_GREATER_EQUAL "8.0.0")
>     add_executable(slh_mitigated microbenchmark.cc)
>     # Compile with -O0 so the array manipulation doesn't get optimized away by the
>     # compiler.
>     target_compile_options(slh_mitigated PUBLIC -mspeculative-load-hardening -O0)
>   endif()
> endif()

Reviewer: ```suggestion
# Speculative load hardening is only available in Clang >= 8.0.0.
# Avoid VERSION_GREATER_EQUAL so we work with CMake < 3.7.
if (("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang") AND
    (NOT "${CMAKE_CXX_COMPILER_VERSION}" VERSION_LESS "8.0.0"))
  add_executable(slh_mitigated microbenchmark.cc)
  # Compile with -O0 so the array manipulation doesn't get optimized away by the
  # compiler.
  target_compile_options(slh_mitigated PUBLIC -mspeculative-load-hardening -O0)
endif()

```

Reviewer: An example of the LLVM folks needing to deal with this: https://reviews.llvm.org/D57326
